### PR TITLE
Feature/v 2 payments

### PR DESCRIPTION
**Pull request by andrii-onufriichuk**

### CODE REVIEW

File: Gateway/Command/CreatePayment.php [hunk 1]
```diff
@@ -48,6 +48,7 @@ public function execute(array $commandSubject)
             'orderId' => $orderId,
             'method' => $method,
             'type' => $type,
+            'captureType' => 'AUTOMATIC_PLUGIN',
```
Comment (aashwin-rvvup) [inline Gateway/Command/CreatePayment.php:51]: Locally, line 57 is overwriting with AUTOMATIC_CHECKOUT, so it's not working as intended. I think we should only overwrite capture type if its not AUTOMATIC_PLUGIN/CHECKOUT

File: Controller/Redirect/In.php [hunk 2]
```diff
@@ -229,34 +249,103 @@ private function redirectToCart()
...
+                    'An error occurred while processing your payment (ID %1). Please contact us. ' . $rvvupId
+                )
+            );
+            return $this->redirectToCart();
         }
+
+        $hash = $quote->getPayment()->getAdditionalInformation('quote_hash');
+        $quote->collectTotals();
+        $savedHash = $this->hashService->getHashForData($quote);
+        if ($hash !== $savedHash || $rvvupId !== $lastTransactionId) {
```
Comment (aashwin-rvvup) [inline Controller/Redirect/In.php:284]: I think we should split this into two different conditions.
First one: rvvupId is not same as lastTransactionId in this case we put a different message saying, the order was already placed please review the transaction before placing again or something along those lines.

Second one: for hash doesnt match, in which case we put the message for cart has changed place another order.

File: Controller/Redirect/In.php [hunk 3]
```diff
@@ -229,34 +249,103 @@ private function redirectToCart()
...
-                $this->messageManager->addWarningMessage(__($processOrderResult->getCustomerMessage()), $messageGroup);
+        if (!$quote->getItems()) {
+            $quote = $this->getQuoteByRvvupId($rvvupId);
+            $lastTransactionId = (string)$quote->getPayment()->getAdditionalInformation('transaction_id');
+        }
+        if (empty($quote->getId())) {
+            $this->logger->error('Missing quote for Rvvup payment', [$rvvupId, $lastTransactionId]);
+            $this->messageManager->addErrorMessage(
+                __(
+                    'An error occurred while processing your payment (ID %1). Please contact us. ' . $rvvupId
```
Comment (aashwin-rvvup) [inline Controller/Redirect/In.php:275]: %1 isn't replaced correctly.

File: Controller/Redirect/In.php (hunk 2 above)
Comment (aashwin-rvvup) [reply Controller/Redirect/In.php:284]: We can also confirm the exact messaging

File: Controller/Redirect/In.php [hunk 4]
```diff
@@ -86,86 +134,62 @@ public function __construct(
...
-
-        if ($rvvupId !== $lastTransactionId) {
+        try {
+            $this->sdkProxy->paymentCapture($lastTransactionId, $rvvupPaymentId);
+            $orderId = $this->quoteManagement->placeOrder($quote->getEntityId(), $payment);
+        } catch (\Exception $e) {
+            $this->sdkProxy->voidPayment($lastTransactionId, $rvvupPaymentId);
             $this->logger->error(
-                'Payment ID when redirecting from Rvvup checkout does not match order in checkout session',
+                'Order placement within rvvup payment failed',
```
Comment (aashwin-rvvup) [inline Controller/Redirect/In.php:172]: We should review this message with something more meaningful for the customer.

File: Controller/Redirect/In.php [hunk 5]
```diff
@@ -197,7 +197,6 @@ public function execute()
             $this->sdkProxy->paymentCapture($lastTransactionId, $rvvupPaymentId);
         } catch (\Exception $e) {
             $this->orderManagement->cancel($orderId);
```
Comment (aashwin-rvvup) [inline Controller/Redirect/In.php:199]: I think we don't do this, incase the rvvup order did capture but there was some other issue where it error internally after capture and we got a error resposne (or a network blip). Then Rvvup would treat it as success send a webhook but the order is cancelled in magento, causing us to fail the update.

Leave it as is, and in this edge case/issues, it just solves it self when the payment is expired etc.

File: Controller/Redirect/In.php [hunk 6]
```diff
@@ -86,86 +133,81 @@ public function __construct(
...
                     'last_transaction_id' => $lastTransactionId,
-                    'rvvup_order_id' => $rvvupId
+                    'rvvup_order_id' => $rvvupId,
+                    'message' => $e->getMessage()
                 ]
             );
             $this->messageManager->addErrorMessage(
                 __(
-                    'An error occurred while processing your payment (ID %1). Please contact us.',
+                    'An error occurred while capturing your order (ID %1). Please contact us.',
```
Comment (jay-rvvup) [inline Controller/Redirect/In.php:199]: `capturing your payment`

I think we should keep the "payment" naming over order when we refer to rvvup objects as we are moving away from order terminology generally.

**Review by jay-rvvup — DISMISSED** [review]

**Review by jay-rvvup — APPROVED** [review]

**Review by jay-rvvup — DISMISSED** [review]

**Review by jay-rvvup — APPROVED** [review]

**Pull request merged**